File: struct.js
class ds {

constructor(newmap, newarr){
    this.map = newmap ?? new Map();
    this.arr = newarr ?? [];
}

    add = function(a) {
        if(this.map.get(a)) return;

        let size = this.arr.length;

        this.map.set(a, size);

        this.arr.push(a);

        return this.arr[this.arr.length-1] === a;
    }

    remove = (a) => {

        if(!this.map.get(a)) return;

        let size = this.arr.length -1;

        let index = this.map.get(a);

        this.map.delete(a);

        let temp = this.arr[size];

        this.arr[index] = this.arr[size];

        this.arr[size] = temp;

        this.arr.pop();

        return this.arr[this.arr.length-1] !== a;

    }

    search = (a) => {

        if(!this.map.get(a)) return;

        let index = this.map.get(a) ?? -1;

        return index ?? -1;

    }

    findRandom = () => {

        let random = Math.floor(Math.random() * this.arr.length-1);
        return this.arr[random];
    }


}

//TEST
let test = new ds();
let i = 30;
while(i-- > 0){
    test.add(i)
}
test.remove(6)
test.search(10);
test.findRandom();

File: struct.java
import java.util.HashMap;
import java.util.Random;
import java.util.ArrayList;
import java.util.Collections;

class dataStruct<T extends Comparable<T>>{
    ArrayList<T> arr;
    HashMap<T, Integer> hashmap;

    public dataStruct() {
        arr = new ArrayList<T>();
        hashmap = new HashMap<T, Integer>();
    }


     void add(T a) {
        if(hashmap.get(a) != null) return;

        int s = arr.size();
        arr.add(a);

        hashmap.put(a, s);
    }

    void remove(T a) {
        Integer index = hashmap.get(a);
        if(index == null) return;

        hashmap.remove(a);

        int s = arr.size();

        T last = arr.get(s -1);

        Collections.swap(arr, index, s-1);

        arr.remove(s-1);

        hashmap.put(last, index);

    }

    T random(){
        Random rand = new Random();
        int index = rand.nextInt(arr.size());

        return arr.get(index);

    }

    Integer search(T a){
        return hashmap.get(a);
    }
}
    


public class struct {

    public static void main (String args[]){
        dataStruct<Integer> ds = new dataStruct<Integer>();

        ds.add(5);
        ds.add(1);
        ds.add(20);
        ds.add(30);
        ds.remove(5);
        System.out.println(ds.random());
        
    }

}